### DISCUSSION TITLE
Only have window shadows with picom when windows are floating?

### DISCUSSION THREAD

**Discussion by alkostuchkov** [qtile/qtile#2944] [Q&A] [answered]
Hello, I'm currently using picom to add shadows to my windows. 
I'd like to add shadows to my windows only if they are floating or if they are a "menu" program like dmenu or rofi.
I wanna add rules to shadow-exclude to my picom.conf, but I don't know how to determine floating windows using xprop.
i3 WM has a matching syntax for picom: "I3_FLOATING_WINDOW@:c" and the solution for i3 is 
"!I3_FLOATING_WINDOW@:c && !class_g = 'Rofi' && !class_g = 'dmenu'". 
(https://www.reddit.com/r/i3wm/comments/j1jxzd/only_have_window_shadows_with_picom_when_windows/) 

Is there the similar solution for qtile?
Thanks in advance.

**Comment by m-col** [comment, marked as answer]
Your best bet is to set a similar window hint yourself using a hook, and then tell picom to watch out for that. For example:
```py
from libqtile import qtile, hook

@hook.subscribe.float_change
@hook.subscribe.client_new
def set_hint(window):
    window.window.set_property("IS_FLOATING", int(window.floating))
```
This will set the hint IS_FLOATING on all managed windows (those menu-like windows like rofi and dmenu are not managed by qtile) that will be 1 if the window is floating or 0 if not.

**Reply by alkostuchkov** [reply]
Thank you for reply.

With your variant:
`window.window.set_property("IS_FLOATING", int(window.floating))`
xprop doesn't show IS_FLOATING at all.

In this [Assigning a WM_CLASS atom to a libqtile.bar.Bar object #2351]() discussion I've found an answer how to see added property in the xprop output:

```
@hook.subscribe.float_change
@hook.subscribe.client_new
def set_hint(window):
    window.window.set_property("IS_FLOATING", str(window.floating), type="STRING", format=8)

```
Now I can see IS_FLOATING in the xprop output.
But xprop always shows IS_FLOATING(STRING) = "False" for all windows (floating and not).

I added the extra line to see window.floating in the notification for client_new and when float_change.

```
@hook.subscribe.float_change
@hook.subscribe.client_new
def set_hint(window):
    window.window.set_property("IS_FLOATING", str(window.floating), type="STRING", format=8)
    subprocess.check_output(["notify-send", "-i", "dialog-information", "{} {}".format(window.name, window.floating)])
```
When new window is created notification is shown but window.floating is false even for those window I added in the float_rules as always floating. And when I change floating mode for a window manually no notification at all.
Why the window.floating is set incorrectly?

**Reply by m-col** [reply]
I just noticed that `float_change` actually doesn't pass the window as an argument. This seems unintuitive to me. Instead you'll have to do something like:
```py
from libqtile import qtile, hook
from libqtile.backend.base import Internal

@hook.subscribe.float_change
def set_hint():
    for window in qtile.windows_map.values():
        if not isinstance(window, Internal):
            window.window.set_property("IS_FLOATING", int(window.floating))

@hook.subscribe.client_new
def set_hint(window):
    window.window.set_property("IS_FLOATING", int(window.floating))
```

I think that might get the float_change hook working.

**Reply by alkostuchkov** [reply]
Unfortunaly not.
Even client_new doesn't set the property correctly. Only after I restart qtile.
I've solved this (thanks you gave me a good idea in your first post).
Solution that works:

```
@hook.subscribe.client_focus
def set_hint(window):
    window.window.set_property("IS_FLOATING", str(window.floating), type="STRING", format=8)
```
In picom.conf I added:
```
shadow-exclude = [
"IS_FLOATING@:s != 'True'"
]
```
Thank you again!

**Reply by m-col** [reply]
Nice, glad you got it working!
